[1단계 - 상품 관리 기능] 루카(백경환) 미션 제출합니다.

dooboocookie opened the pull request:
안녕하세요 카일!! 🐈🐈‍⬛
백엔드 5기 루카🐟입니다!!

## 중요하게 생각한 부분
  - CRUD API를 만들면서, `요청 메소드`를 어떤 것을 해야되고, `응답 상태 코드`어떤 것을 넘겨야 되고, `예외 핸들링`을 어떻게 할지에 대해서 집중해보았습니다.
  - Controller - Service - DAO 각각의 `단위테스트`와 전체적인`통합?인수? 테스트`를 어떻게 해야될지 집중해보았습니다.


## 질문사항
### 1.  Entity와 Domain을 동일시?
- 저희는 처음에 Domain자체를 만들지 않을 생각이였습니다.
- 현재로서는 도메인에게 필요한 로직이 검증로직 밖에 없었습니다.
- Entity는 DAO(Persistence Layer)에서 DB 테이블과 1:1로 매핑되기 위한 객체로 만들었습니다.
- Entity를 만들고 보니, 해당 Entity가 도메인으로 사용되도 괜찮을 것 같은 생각이 들었습니다.
  - (필요없는 id 필드가 있지만)
- 저번 미션에서는 도메인 객체랑 엔티티 객체를 따로 두었는데, 도메인 객체가 어떤 동작을 한 후 그 상태를 엔티티 객체와 매핑해서 데이터를 영속화 해야되는데, 그 부분에 어려움이 있었습니다.
- 그래서 이번에는 엔티티 자체를 도메인으로서 사용하려고 하는데, 이 부분에 대해서 어떻게 생각하시는지 궁금합니다.

### 2. PUT과 DELETE의 응답 코드 & 응답 메세지는?
- CRUD API를 만들면서, `/items`라는 요청에 대해서 HTTP 요청 메소드를
  - create : POST 요청 -> 201 CREATED / content-x
  - read : GET 요청 -> 200 OK / content-상품리스트 or 상품(JSON)
  - update : PUT 요청 ->  201 CREATED / content-x
  - delete : DELETE 요청 ->  200 OK / content-x
- create, update, delete 를 하면서 딱히 응답에 넣어 줘야할 content가 없다고 느껴집니다.
  - create, update, delete 같은 처리를 하고 나서는 응답에 content 어떤것을 줘야할지 궁금합니다.
  - 그리고 만약에 진짜 넣어줘야할 content가 없다면, 204 NO CONTENT 로 응답하는 것은 어떤지 궁금합니다.
- update, delete를 하고 나서는 어떤 상태코드를 응답해야되는지 궁금합니다.
  - update는 원래 있던 자원을 다른 값으로 변경하는 것이므로 201 CREATED로 하였고,
  - deleted는 새로운 자원을 만드는 것은 아니므로 200 OK로 하였습니다.

### 3. 단위 테스트 시 모킹👌? 통합 테스트 시 모킹🙅?
- 저희는 이번에 Controller-Service-DAO에 대한 단위테스트와 `요청` &rarr; `응답`과정을 테스트하는 테스트를 하였습니다.
- 각 단위 테스트에서는 
  - Mocking을 최대한 활용하여, 지금 테스트하고자하는 레이어에 해당되는 부분만 테스트하였습니다.
  - 그렇다 보니, 저희가 모킹한대로만 결과를 내뱉는 듯한 느낌이 들어서 테스트가 약간 무의미하다는 생각이 들었습니다.
  - 그렇다고, DaoTest에서 이미 한 테스트를 하고 잘 작동한다는 것이 확실하니 ServiceTest에선 Dao를 Mocking하여 사용하였는데, 이것에 대해서 어떻게 생각하시는지 궁금합니다.
- `요청` &rarr; `응답`에 대한 테스트(JwpCartApplicationTests)에서는
  - 어떤 요청을 보냈을 때, 어떤 응답이 오는지에 대한 테스트를 진행하였는데, 모킹을 전혀 사용하지 않고 테스트를 하였습니다.
  - DB에 저장되는 내용까지 고려하여 테스트를 진행해야 됐습니다.
    - 응답하는 내용이 DB에 저장되어있는 값에 의존적인데,
    - 현재는 매번 DB를 초기화 시켜서 그 값을 알기 쉽지만, 만약 내용이 커진다면 어떤 응답을 테스트해야할지 궁금합니다.
 
### 4. 검증 위치
- 현재 저희는 검증을 2군데서 진행합니다. `Controller에서 DTO 바인딩 시`, `도메인(엔티티)에 필드 값 설정 시`
- 동일한 내용을 두 군데서 하기도 합니다.
- 저희는 각 영역이 갖는 검증의 목적이 다르다고 생각하였습니다.
  - Controller: 좋은 요청을 받기 위해
  - Domain: 도메인이 무결하기 위해
- 이런식으로 동일한 내용을 두 군데서 검증하는 것에 대해 어떻게 생각하시는지 궁금합니다.

---

아직 부족함이 많다보니, 질문을 좀 장황하게 하게 되었습니다 😭
바쁘신 와중에 긴 글 읽어주셔서 감사합니다!!

리뷰 잘 부탁드리겠습니다🙇‍♂️🙇‍♂️🙇‍♂️

JJ503 committed:
docs: 기능 요구사항 목록 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: Datasource 등록 및 item 테이블 생성

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: item 전체 조회 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: item 전체 조회 기능 테스트

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemService에서 전체 아이템 불러오기 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ListController에 GET / 요청 매핑

- index.html에 상품 정보에 맞게 thymeleaf 변경

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: AdminController GET /admin 요청 매핑

- admin.html에 상품 정보에 맞게 thymeleaf 변경

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemDAO 저장하는 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
refactor: item 빌더 패턴 적용

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemService 아이템 저장 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemController에 POST /items 요청 매핑

- admin.html 상품등록 모달창 input 태그 name속성 지정
- admin.js axios 요청 설정

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemController에 GET /items 요청 매핑

- item 목록을 응답한다.

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemDao 수정 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemService 수정 기능 추가

- 요청 DTO 추가
- 저장 DTO 네이밍 변경

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemController PUT /items/{id} 요청 매핑

- {id}의 정보를 수정하는 기능을 한다.
- admin.js axios 요청 설정 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemDao 삭제 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemService 삭제 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemController DELTE /items/{id} 요청 매핑 추가

- {id} 정보를 삭제하는 기능을 한다.
- admin.js axios 요청 설정 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemController DELTE /items/{id} 요청 매핑 추가

- {id} 정보를 삭제하는 기능을 한다.
- admin.js axios 요청 설정 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
Merge remote-tracking branch 'origin/step1' into step1

JJ503 committed:
docs: 완료한 요구사항 목록 체크

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: GET /items/{id} 요청 매핑

- 하나의 아이템 조회 기능 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: item 등록, 수정, 삭제 시 메인 페이지로 이동하도록 수정

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: 모든 요청에 대한 통합 테스트

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: ItemDao 단위 테스트 추가

- initializeTestDb.sql로 테스트마다 테이블 DROP 후 CREATE하도록 변경

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: Item의 필드를 랩핑 후 각 필드의 검증 로직 추가

- Item을 도메인으로 고려
- ItemResponse DTO를 생성

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: ItemService 테스트 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
refactor: ItemDao findBy() optional 반환으로 변경

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: ItemRequest의 검증과 예외처리 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
chore: 필요없는 파일 삭제

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
feat: 도메인 검증 로직 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: 도메인 생성 및 예외 테스트 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: Dao 생성 및 예외 테스트 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: 서비스 생성 및 예외 테스트 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: 아이템 컨트롤러 테스트 추가

- 요청과 컨트롤러 메서드 매핑 테스트
- 컨트롤러 메서드 반환값 테스트

Co-authored-by: dooboocookie <bkh3030@gmail.com>

JJ503 committed:
test: ExceptionHandler 응답 테스트 추가

Co-authored-by: dooboocookie <bkh3030@gmail.com>

KIMSIYOUNG commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final ItemService itemService;
+
+    @Autowired
```
생성자 주입시에는 해당 어노테이션을 사용하지 않아도 주입이 되어요 🙂

dooboocookie replied:
아 넵 제거하겠습니다!
생성자가 1개 있을 때 @Autowired를 생략해도 된다는 것을 학습한 적이 있습니다.
그럼에도 어노테이션을 붙인 나름대로의 이유는 DI를 명시적으로 나타내고 싶어서 붙였습니다!!

혹시 @Autowired를 생략하는 것이 어떤 컨벤션에 해당이되거나 좋은 점이 있을까요? 아니면 굳이 붙이지 않는 느낌일까요?

KIMSIYOUNG replied:
1. 불필요한 스프링 의존성을 드러낼 필요는 없을 것 같아요.
2. 말씀하신대로 의미를 명확하게 해주는 효과가 있다고 볼 수도 있을 것 같아요. 다만 스프링 환경에서 개발중이라면 대부분 다 알고 있는 것이라 생략하는게 더 자연스러울 것 같아요.

KIMSIYOUNG commented on src/main/java/cart/controller/ExceptionItemControllerAdvice.java:
```diff
@@ -0,0 +1,59 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+// TODO: 2023/04/26 text/html을 기대할 떄와, application/json을 기대할 떄 똑같이 Json으로 예외 결과를 응답하는게 맞을까?
+@ControllerAdvice
```
`@RestControllerAdvice`와는 어떻게 다른 어노테이션인가요?

dooboocookie replied:
```java
 * A convenience annotation that is itself annotated with
 * {@link ControllerAdvice @ControllerAdvice}
 * and {@link ResponseBody @ResponseBody}.
 * ...
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@ControllerAdvice
@ResponseBody
public @interface RestControllerAdvice {
  //...
}
```
@Conroller와 @RestController의 차이와 마찬가지로 
@ControllerAdvice에 @ResponseBody어노테이션이 추가된 것이라고 생각하고 있습니다.

만약 해당 ControllerAdvice에서 핸들링하는 에러의 정보를 뷰페이지가 아니라 application/json으로 주기로 정했다면 @RestControllerAdvice로 하는 것이 맞는것 같습니다.

todo 주석에 대해서 생각이 정리 되었습니다.
### 1. ItemController의 어노테이션을 `@Controller` &rarr; `@RestController`
-  `/items`의 요청과 매핑되는 ItemController는 `item`리소스의 CRUD하는 역할
- @RestController가 적절
### 2. 이 ControllerAdvice의 어노테이션을 `@ControllerAdvice` &rarr; `RestControllerAdvice` 
- 이 ControllerAdvice의 범위를ItemController로 변경
- 담당하는 컨트롤러가 `@RestController`이므로 요청의 accept혹은 응답의 content-type이 `application/json`으로 예상
### 3. `/` 요청과 `/admin`과 같이 `text/html`을 응답하는 컨트롤러에 대한 ControllerAdvice를 따로 지정
- 거기서 발생되는 예외에 응답을 줄 만한 예외 페이지 생성

KIMSIYOUNG replied:
잘 학습해주셨네요🙂 말씀하신 기준대로 적용하면 될 것 같습니다.

KIMSIYOUNG commented on src/main/java/cart/controller/ExceptionItemControllerAdvice.java:
```diff
@@ -0,0 +1,59 @@
+                             .body(exceptionResponse);
+    }
+
+    @ExceptionHandler(NotFoundResultException.class)
```
1. `IllegalArgumentException`과 동일한 처리를 하고 있는데 별도로 선언하신 이유는 무엇일까요?
2. 커스텀 예외로 정의하신 이유는 무엇인가요?

dooboocookie replied:
> 1. IllegalArgumentException과 동일한 처리를 하고 있는데 별도로 선언하신 이유는 무엇일까요?

큰 고려 없이, 각 예외마다 다른 응답이 있을 수 있다 생각하여 각 `IllegalArgumentException`, `MethodArgumentNotValidException`, `NotFoundResultException`를 따로 핸들링 하였는데, 결과적으론 다 같은 응답을 주고 있습니다...
일단 같은 동작은 하는 예외는 합치도록 하였습니다.

> 2. 커스텀 예외로 정의하신 이유는 무엇인가요?

1번과 같은 이유로 저 `수정이나 삭제를 하려고 시도한 상품이 없을 때` 발생하는 에러를 특정해서 그에 대한 특별한 응답을 줘야겠다는 생각이 들었습니다.
일단`NotFoundResultException`는 사용하지 않고, `IllegalArgumentException`하나로 처리하도록 하겠습니다.

KIMSIYOUNG replied:
네네!
두 질문 모두 불필요하다는 의미는 아니었어요. 
따로 잡거나, 커스텀 예외를 사용하는 것 자체는 좋은 것 같습니다. 다만 다른 것들과 동일한 처리라면 `굳이` 그렇게 처리할 필요가 없을 것 같다는 의미였어요.
다른 예외핸들링을 하기 위해서라면 별도로 커스텀 예외를 정의하고 이를 잡아서 처리하는 것도 좋을 것 같아요.

KIMSIYOUNG commented on src/main/java/cart/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @PostMapping
+    public ResponseEntity addItem(@RequestBody @Validated final ItemRequest itemRequest) {
```
1. 제네릭으로도 Void를 사용할 수 있습니다.🙂
2. Location 을 내려주는 이유는 무엇인가요?

dooboocookie replied:
> 1.제네릭으로도 Void를 사용할 수 있습니다.🙂

아 그렇군요! ㅠㅠ 
응답이 없을 때, `ResponseEntity<Void>`로 반환타입 지정하겠습니다.

> 2. Location 을 내려주는 이유는 무엇인가요?

미션의 초기 상태는 axios요청에서 정상 응답이 오면 페이지를 reload하도록 되어있었습니다. 
`C(저장)`, `U(수정)` 후에는 새로고침보단 상품에 대한 페이지로 가야된다고 생각했는데, 해당 미션에는 해당 페이지는 없어서 메인페이지로 가는 `201상태`와 `상품목록 페이지로 가는 location 헤더`를 응답으로 보냈습니다.
`D(삭제)`후에는 위와 상품에 대한 페이지가 애초에 없을 것이여서 `200상태`와 `상품목록 페이지로 가는 location 헤더`를 응답으로 주었습니다.

근데 제가 찾아봤을 때 location 해더는 보통 redirect응답이나, 201응답에서 추가하는 헤더라고 하더라구요. 그래서 삭제 시에도 주는게 맞을지는 잘 모르겠습니다.

KIMSIYOUNG replied:
일반적으로 어떠한 상황에서 사용되는지 잘 인지하고 계신 것 같아요. 
그렇기 때문에 삭제요청에서 이를 다루는 것은 문제되지 않을 것 같습니다.
요청 응답 포맷은 기본적으로 모두가 동의하는 값을 사용하되, 해석의 여지가 달라질 수 있는 부분은 협업하는 곳과 함께 정의하고 그 정의에 따라 사용하면 될 것 같아요.
전체적으로 어떤 상태코드와 어떤 조합이 사용되는지에 대해만 인지하고, 실제 값은 협업을 통해서 결정하면 될 것 같습니다!

KIMSIYOUNG commented on src/main/java/cart/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+
+    @PutMapping("/{itemId}")
+    public ResponseEntity updateItem(@PathVariable final Long itemId,
+                                     @RequestBody @Validated final ItemRequest itemRequest) {
```
`@Validated`와 `@Valid`는 어떻게 다른가요?

dooboocookie replied:
이 부분 또한, 크게 의식하지 않고 `@Validated`가 `@Valid`의 기능을 포함하고 있다고 알고 있어서 `@Validated`로 적용하였습니다.

다시 찾아보니, `@Valid`는 JSR-303를 기반으로 하는 자바 표준 `import javax.validation.Valid`이고,
`@Validated`는 스프링프레임워크에서 제공하고 AOP기반으로 작동하는 `import org.springframework.validation.annotation.Validated`로 확인했습니다.

아직은 둘을 써야하는 상황에 대해서 확신은 없는 상태긴 하지만

```java
import org.hibernate.validator.constraints.Length;
import org.hibernate.validator.constraints.URL;

import javax.validation.constraints.Max;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Positive;

public class ItemRequest {
  // ...
}
```
인 상태로로 Controller에서 `@Valid`로도 충분히 가능한 것 같아서 `@Valid`로 변경하겠습니다.

KIMSIYOUNG replied:
네네!
Valid는 입력값 검증에 주로 사용되고, Validated는 메소드파라미터를 검증할 때 주로 사용하는 것 같아요.
다만 말씀하신대로 두가지가 모두 처리할 수 있는 영역에선 어떤 것을 써야하는지 모호할 수 있을 것 같아요.
저는 굳이 스프링에 조금 더 의존할 필요는 없을 것 같아 Valid를 주로 사용하는 것 같습니다.

KIMSIYOUNG commented on src/main/java/cart/service/ItemService.java:
```diff
@@ -0,0 +1,71 @@
+
+    public ItemResponse loadItem(final Long itemId) {
+        Optional<Item> findItem = itemDao.findBy(itemId);
+        return ItemResponse.from(findItem.orElseThrow());
```
orElseThrow()를 이렇게 사용해도 괜찮을까요?

dooboocookie replied:
앗 안될 것 같습니다.

수정 삭제를 했을 때, 대상 상품이 없으면 예외를 발생시킨 것 처럼,
이 경우에도 IllegalArgumentException을 발생 시켜 핸들링해야될 것 같습니다!

KIMSIYOUNG commented on src/main/java/cart/service/ItemService.java:
```diff
@@ -0,0 +1,71 @@
+        return itemDao.save(item);
+    }
+
+    public void updateItem(final Long itemId, final ItemRequest itemRequest) {
```
업데이트와 삭제를 하는 방법에는 두가지가 있을 것 같아요.

1. 현재와 같이 찾아와서 확인하고 없다면 예외 있다면 로직 수행
2. 있는 것과 무관하게 삭제 또는 업데이트 수행(없었다면 실제 데이터의 변경은 아무것도 일어나지 않겠죠)

두가지 중 첫번째로 작성해주신 이유가 있으실까요?

dooboocookie replied:
`2`의 경우 아무일도 일어나지 않고 SQL예외도 발생되지 않기 때문에
`아무 일도 없었다.`라는 상황을 어플리케이션도 클라이언트도 바로 알아차리지 못했을 것이라 생각했습니다.

그래서 `1`로 먼저 있는지 조회를 하고, 없으면 그 변경(수정, 삭제)이 `BAD REQUEST`라고 응답을 줄 수 있게하였습니다.

KIMSIYOUNG commented on src/main/java/cart/service/ItemService.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.stream.Collectors;
+
+@Service
+public class ItemService {
```
트랜잭션에 대해서도 학습해보시면 좋을 것 같아요.

dooboocookie replied:
앗 넵!! 한번 알아보고 적용하였는데요. 더 깊게 학습하겠습니다.

지금은 조회하는 작업에는 `@Transactional(readonly=true)` 삽입,수정,삭제 작업에는 `@Transactional`을 적용한 상태입니다.

KIMSIYOUNG commented on src/main/java/cart/domain/Item.java:
```diff
@@ -0,0 +1,100 @@
+
+import java.util.Objects;
+
+public class Item {
```
말씀하신대로 도메인과 엔티티는 다를 수 있고 같을 수도 있습니다. 
그렇기에 도메인과 엔티티를 분리하는 것의 장점을 누릴 수 있다면 분리를 하는 것이고, 그렇지 않다면 분리하지 않아도 된다고 생각합니다.
루카가 느끼는 장단점은 무엇인가요?

dooboocookie replied:
## 제가 생각하는 장단점

### `도메인` &harr; `엔티티`를 동일시할 때 장점
- `도메인 객체 == 엔티티 객체`를 영속화 하는데 장점이 있을 것 같습니다.
- 도메인 로직을 돌리고 그 데이터를 그대로 영속화 하는 과정에서 다시 엔티티로 변환을하거나 하는 과정이 없으므로,
- 서비스에서 도메인 로직을 호출하고 그를 그대로 영속화하기 위해 DAO를 호출하는 과정이 간단하게 진행될 수 있을 것 같습니다.

###   `도메인` &harr; `엔티티`를 분리할 때 장점
- DB의 영역과 도메인의 영역을 확실하게 구분할 수 있을 것 같습니다.
- 저번 자동차 미션에서는 `Domain 영역`을 레벨 1때 작성하고, 
- `DB`와 그에 대한 `DAO`, `Entity`를 레벨 2때 작성하였는데요. 
- 그렇다보니, DB의 변경이 도메인의 영향을 끼치거나, 도메인의 변경이 DB의 영향을 끼치지 않는다는 장점이 있을 것 같습니다.

KIMSIYOUNG replied:
장단점을 잘 이해하고 계신 것 같아요.
프로젝트 규모에 따라 또는, 성격에 따라 둘을 분리하지 않아도 되는 경우도 있는 것 같아요.
그렇기에 둘을 분리했을 때 발생하는 중복코드 및 변환 코드의 비효율보다, 얻는 장점이 많다면 분리하는게 좋고 그렇지 않다면 분리하지 않아도 괜찮을 것 같습니다.

KIMSIYOUNG commented on src/main/java/cart/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+    @DeleteMapping("/{itemId}")
+    public ResponseEntity deleteItem(@PathVariable final Long itemId) {
+        itemService.deleteItem(itemId);
+        return ResponseEntity.status(HttpStatus.OK)
```
상태 코드는 말씀하신대로 `해석`에 따라 달라지고 지켜야만 하는 `법칙`이 아니라 이렇게 하자고 만든 `규칙`에 가까운 것 같아요.
그렇기에 200번대 400번대 500번대는 구분되어야 하지만 그 내외에서 사용되는 상태코드값은 합의된 값이라면 어떤 값이라도 문제되지 않는 것 같습니다.
다만 현재는 학습차원이기에 신뢰성있는 곳에서 가이드하는 코드값을 익혀보고 사용하는 것이 좋을 것 같습니다.
가이드에서 모호한 부분이 있다면 그 부분은 루카가 해석한대로 하되, 함께 작업하는 분 + 이 API를 사용하는 분들과 협의하면 될 것 같아요.

dooboocookie replied:
좋은 말씀 감사합니다!!

현재까지 제가 페어와 해석한 것은 `DELETE`의 경우 해당 요청에 대한 처리를 완전히 끝낸 것이므로 `200응답`으로 하기로하였습니다.
일단은 이 합의대로 가고, 더 학습하다가 바꿀 필요가 있으면 그때 팀이나 사용하는 사람과 혐의를 통해 바꾸도록하겠습니다!

KIMSIYOUNG commented on src/main/java/cart/controller/dto/ItemRequest.java:
```diff
@@ -0,0 +1,39 @@
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Positive;
+
+public class ItemRequest {
```
1. 수정과 생성에서 동일한 요청을 사용하는건 어떤 장단점이 있을까요?
2. 두 곳에서 동일한 검증을 한다는 것은 서로의 영역에 서로가 침범되어 있다고 표현할 수도 있을 것 같습니다. 각 레이어별로 검증해야하는 역할은 다른데 이를 모든 곳에 넣어둔다면, 도메인 검증이 수정되어도 입력에 수정되고, 입력이 수정되어도 도메인이 수정될 것 같아요.

dooboocookie replied:
> 1. 수정과 생성에서 동일한 요청을 사용하는건 어떤 장단점이 있을까요?
- 일단 지금은 `생성`과 `수정` 요청에서 입력받는 데이터가 같고, 검증해야될 것도 같다고 생각하여 하나의 `ItemRequest`로 입력받았습니다.
- 하나의 데이터 형식으로 두 입력을 관리하니까 코드의 중복을 줄일 수 있고,
- 반면, 수정이나 생성이 서로를 침범할 수 있는 것이 단점 같습니다.
- 그리고, 수정 시 여러 필드 중 이름만 바꾸고 싶어도 이름의 정보만 입력하는 것이 아니라, 전체 데이터를 입력할 떄 처럼 똑같이 유효성검사에 맞춰 다 보내야된 것이 단점 같습니다.

> 2. 두 곳에서 동일한 검증을 한다는 것은 서로의 영역에 서로가 침범되어 있다고 표현할 수도 있을 것 같습니다. 각 레이어별로 검증해야하는 역할은 다른데 이를 모든 곳에 넣어둔다면, 도메인 검증이 수정되어도 입력에 수정되고, 입력이 수정되어도 도메인이 수정될 것 같아요.

- `도메인`과 `입력` 두 군데 같은 검증을 둔 이유는
  - `도메인`은 어떤 입력이 들어오건 어떤 컨트롤러와 연결되건 별개로 이 도메인 로직이 온전하게 돌아가기 위한 검증이였습니다.
  - `입력`은 요청이 올 떄 그 요청에서 입력된 값이 적절한 값인지 도메인까지 가지 않더라도 입력차원에서 거르기 위함이였습니다.
- 그렇다 보니, 현재로서는 같은 검증 로직이 두군데 있는 것은 어쩔 수 없다는 생각이 드는데,
- 만약 한군데만 꼭 검증이 있어야한다면 그것은 `도메인`같습니다.
  - 여러 컨트롤러와 연결되더라도 도메인이 완전한 검증을 한다면 적절한 예외 응담을 줄 수 있을 것 같습니다.

KIMSIYOUNG replied:
1. 네네 단점에 대해서 잘 인지하고 계시는 것 같습니다. 말씀하신 단점으로 버그가 발생할 수도 있어요. 그렇기에 중복 코드가 발생하더라도 미리 분리해둘지, 함께 사용하되 변경이 생긴다면 2곳에서 영향을 받는다는 것을 다른 개발자가 인지할 수 있게 장치가 필요할 것 같아요.
2. 네네 어플리케이션에서 검증은 매우 중요하지만, 검증에 대한 불안감으로 이곳저곳에서 동일한 검증을 수행하면 오히려 코드 가독에 어려움이 생기고, 서로의 역할에 침범함으로써 변경포인트를 제대로 인지하지 못하는 단점도 있는 것 같아요. 그렇기에 입력에서는 최소한의 값검증만 수행하고 도메인 규칙은 도메인에서 검증하는게 좋을 것 같습니다.

KIMSIYOUNG commented on src/test/java/cart/controller/ItemControllerTest.java:
```diff
@@ -0,0 +1,210 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+@SpringBootTest
```
의존하는 곳을 목처리한다면 스프링부트 테스트로 수행하지 않아도 될 것 같아요.

dooboocookie replied:
```java
class ItemControllerTest {

    @MockBean
    ItemService mockItemService;

    @MockBean
    ItemController mockController;

    @Mock
    ItemService itemService;

    @InjectMocks
    ItemController itemController;

    @Autowired
    MockMvc mockMvc;
    // ...
}
```
해당 테스트에서 2가지를 테스트하기로 결정했었습니다.
1. 컨트롤러 메서드가 어떤 요청과 잘 매핑 되어있는지
2. 컨트롤러 메서드가 적절한 ResponseEntity를 반환하는지

그래서 `@ExtendWith(MockitoExtension.class)`만으로는 테스트가 불가능했습니다.
현재는 `@WebMvcTest`로 변경해놓은 상태입니다.

KIMSIYOUNG replied:
넵넵! 스프링부트테스트가 아니라, 딱 필요한 의존성만 가지고 뜰 수 있게 잘 수정해주셨네요.

KIMSIYOUNG commented on src/test/java/cart/JwpCartApplicationTests.java:
```diff
@@ -1,13 +1,202 @@
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc
 class JwpCartApplicationTests {
```
요기서 질문주신 3-2번쨰 내용이 이해가 잘 안되어요! 조금 더 자세히 설명해주실 수 있나요?

dooboocookie replied:
`요청` &rarr; `응답`에 대한 테스트하는 과정에서 이미 DB에 있는 값을 알아야 테스트가 어떤 응답이 나올지 아는 상황입니다.
그래서 각 테스트마다`@Transactional`로 롤백이 된다해도 
DB에 있는 값을 알기 위해서 `@Sql("classpath:initializeTestDb.sql")`이런식으로 테스트할 때마다 DB를 원하는대로 초기화 시킵니다.
실제 서비스를 하는 프로덕트에서는 이런식으로 DB를 초기화는 못시킬 것 같은데,
이런 류의 테스트를 할 때, 이 테스트 코드 하나만으로 `요청` &rarr; `응답`의 시나리오가 잘 진행된다라고 확인할 수 있는 테스트는 어떻게 짜는게 좋을지 아직 고민입니다..

KIMSIYOUNG replied:
프로덕트환경에서도 테스트는 로컬환경에서 띄우는 방식으로 데이터를 미리 세팅할 수 있습니다.
그렇기에 크게 문제되지 않을 것 같아요. 
추후에 프로파일 분리, 환경분리 등으로 테스트에서만 사용할 수 있는 데이터 세팅을 할 수 있을 것 같아요.

KIMSIYOUNG commented on src/test/java/cart/service/ItemServiceTest.java:
```diff
@@ -0,0 +1,148 @@
+
+// TODO: 2023/04/26 서비스 테스트가 할게 없음 -> Mockito를 쓰는게 맞을까?
+@ExtendWith(MockitoExtension.class)
+class ItemServiceTest {
```
말씀하신대로 현재의 로직에서는 서비스의 역할은 단순히 Dao에 값을 전달하는 역할밖에 없으니 당연한 것 같습니다.
다만 비즈니스 로직이 전개된다면 다오에 대한 의존으로 인해 테스트가 꺠지는 것인지, 서비스 로직이 문제가 되는지 확인하기 어려워질 수 있을 것 같아요.
목테스트를 사용하지 않았을 때 얻게되는 장점은 무엇인가요?

dooboocookie replied:
앗 그렇군요..
현재 Service에서 하는 역할이 크지 않아서 그렇게 느낀 것 같습니다.
지금 테스트를 하고싶은 포인트가 Service에 있다면 의존하는 Dao는 Mocking을 하여 Dao가 정상적으로 잘 동작한다고 가정하고, Service에 대해서만 테스트를 하는 것이 맞는것 같습니다.

KIMSIYOUNG reviewed (changes requested):
안녕하세요 루카!
1단계 잘 구현해주셨네요.
몇가지 코멘트 남겼는데 확인 부탁드립니다.
질문 내용은 코멘트에 같이 달아두었어요!

dooboocookie committed:
chore: 생성자 1개 컴포넌트들의 @Autowired 생략

dooboocookie committed:
refactor: RestController와 Controller 구분

- item CRUD 하는 Controller를 @RestController로 변경
- 그에 대한 ControllerAdvice를 @RestControllerAdvice로 변경
- 상품 목록과 관리자 페이지 보여지는 Controller로 한곳으로 모음
- 그에 대한 ControllerAdvice 생성
- 예외 페이지 추가

dooboocookie committed:
refactor: Custom예외 삭제

dooboocookie committed:
chore: 로타입 제거

dooboocookie committed:
refactor: Controller 검증 @Validated에서 @Valid로 변경

- imageUrl 필드에 @URL 검사 추가

dooboocookie committed:
feat: ItemService에 @Transactional 추가

dooboocookie committed:
chore: ResponseEntity<Void> 테스트에 적용 못했던 것 추가

dooboocookie committed:
chore: 찾은 아이템이 없을 시 IllegalArgumentException 발생하도록 변경

dooboocookie committed:
chore: ItemControllerTest @SpringBootTest에서 @WebMvcTest로 변경

dooboocookie commented:
안녕하세요! 카일!!!😸

꼼꼼한 피드백 정말 감사합니다.

제가 궁금해 했던 내용 뿐 아니라 페어랑 코딩 하면서 의식적으로 고민하지 못했던 부분들 까지도 짚어 주셔서 공부할 키워드를 많이 가져갈 수 있었던 것 같습니다.

아직 모르는 내용이 많고 100% 제 기준을 잡진 못했지만, 

최대한 수정할 건 수정하고, 제 원래 코드에 대한 의견을 견고하게 할 수 있는 시간이였습니다!!

감사합니다!!🙇‍♂️

KIMSIYOUNG reviewed (approved):
안녕하세요 루카!
리뷰 반영 잘 해주셨네요.
1단계는 여기서 머지하면 좋을 것 같아요.
궁금하신 내용이나 조금 더 이야기 나누고 싶은 내용은 2단계에 함께 말씀해주시면 좋을 것 같습니다.
고생하셨습니다.
